pkg/scheduler/actions/allocate/allocate.go

```diff
@@ -164,123 +156,146 @@ func (alloc *Action) Execute(ssn *framework.Session) {
 		klog.V(3).Infof("Try to allocate resource to %d tasks of Job <%v/%v>",
 			tasks.Len(), job.Namespace, job.Name)
 
-		stmt := framework.NewStatement(ssn)
-		ph := util.NewPredicateHelper()
-		for !tasks.Empty() {
-			task := tasks.Pop().(*api.TaskInfo)
+		alloc.processTasks(tasks, job, jobs, queue, ssn, allNodes)
+	}
+}
 
-			if !ssn.Allocatable(queue, task) {
-				klog.V(3).Infof("Queue <%s> is overused when considering task <%s>, ignore it.", queue.Name, task.Name)
-				continue
-			}
+func (alloc *Action) processTasks(tasks *util.PriorityQueue, job *api.JobInfo, jobs *util.PriorityQueue, queue *api.QueueInfo, ssn *framework.Session, allNodes []*api.NodeInfo) {
+	stmt := framework.NewStatement(ssn)
+	ph := util.NewPredicateHelper()
 
-			klog.V(3).Infof("There are <%d> nodes for Job <%v/%v>", len(ssn.Nodes), job.Namespace, job.Name)
+	predicateFn := func(task *api.TaskInfo, node *api.NodeInfo) ([]*api.Status, error) {
```

Comment:
Before this modification, predicateFn just declared once, it's declared repeatedly now, we should keep declare it once.

Reply:
yeah